gap/prop.gi

```diff
@@ -567,3 +567,118 @@ function(gr)
+# 2. (out degree of y) + (in degree of x) ≥ n
+# Is Hamiltonian.
+InstallMethod(IsHamiltonianDigraph, "for a digraph",
+[IsDigraph],
+function(gr)
+  local out, indegs, fulldegs, outdegs, n, checkMT, check41, check42, iter,
+        dominatedcheck, dominatingcheck, adjmatrix, i, j, k, tempblist;
+
+  if DigraphNrVertices(gr) <= 1 and IsEmptyDigraph(gr) then
+    return true;
+  elif not IsStronglyConnectedDigraph(gr) then
+    return false;
+  fi;
+
+  if IsMultiDigraph(gr) then
+    gr := DigraphRemoveAllMultipleEdges(gr);
+  fi;
+  if DigraphHasLoops(gr) then
+    gr := DigraphRemoveLoops(gr);
+  fi;
+
+  n := DigraphNrVertices(gr);
+
+  if n <= 512 then
+    out := OutNeighbours(gr);
+    indegs := InDegrees(gr);
+    outdegs := OutDegrees(gr);
+    fulldegs := indegs + outdegs;
+    adjmatrix := BooleanAdjacencyMatrix(gr);
+    #checks if Meyniel's theorem, Theorem 4.1 or Theorem 4.2 are applicable.
+    checkMT := true;
+    check41 := true;
+    check42 := true;
+    for i in [1 .. n] do
+       for j in [1 .. n] do
+          if i <> j and not adjmatrix[j][i] and not adjmatrix[i][j] then
+              #Meyniel's theorem
+              if checkMT and fulldegs[i] + fulldegs[j] < 2 * n - 1 then
+                  checkMT := false;
+              fi;
+
+              if check41 or check42 then
+                dominatedcheck := false;
+                dominatingcheck := false;
+                for k in [1 .. n] do
+                  if adjmatrix[k][i] and adjmatrix[k][j] then
+                     dominatedcheck := true;
+                     break;
+                  fi;
+                od;
+                tempblist := adjmatrix[i];
+                IntersectBlist(tempblist, adjmatrix[j]);
+                dominatingcheck := true in tempblist;
+              fi;
+
+              #Theorem 4.1
+              if check41 and dominatedcheck then
+                 if fulldegs[i] < n - 1 or fulldegs[j] < n - 1 or
+                    (fulldegs[i] = n - 1 and fulldegs[j] = n - 1) then
+                   check41 := false;
+                 fi;
+              fi;
+
+              #Theorem 4.2
+              if check42 and (dominatingcheck or dominatedcheck) then
+                 if (indegs[i] + outdegs[j]) < n or
+                    (indegs[j] + outdegs[i]) < n then
+                   check42 := false;
+                 fi;
+              fi;
+          fi;
+          if not (checkMT or check41 or check42) then
+            break;
+          fi;
+      od;
+      if not (checkMT or check41 or check42) then
+        break;
+      fi;
+    od;
+    if checkMT or check41 or check42 then
+      return true;
+    fi;
+
+    if n < 256 then
+      return DigraphMonomorphism(CycleDigraph(n), gr) <> fail;
+    fi;
+  fi;
+
+  iter := IteratorOfPaths(gr, 1, 1);
+  n := n + 1;
+  while not IsDoneIterator(iter) do
+    if Length(NextIterator(iter)[1]) = n then
+      return true;
+    fi;
+  od;
+  return false;
+end);
```

Comment:
After this, you should add the following method:
```
InstallMethod(IsHamiltonianDigraph, "for a digraph with hamiltonian path",
[IsDigraph and HasHamiltonianPath],
x -> HamiltonianPath(x) <> fail);
```
Because, if we've already calculated a path, we already know the answer.

Reply:
And also, add a test for this case.